javapoet/src/main/java/com/palantir/javapoet/TypeSpec.java

```diff
@@ -299,6 +307,11 @@ void emit(CodeWriter codeWriter, String enumName, Set<Modifier> implicitModifier
                 if (kind == Kind.INTERFACE) {
                     extendsTypes = superinterfaces;
                     implementsTypes = Collections.emptyList();
+                } else if (kind == Kind.RECORD) {
+                    extendsTypes = Collections.emptyList();
+                    implementsTypes = superinterfaces;
+                    // Record constructor
+                    emitRecordConstructor(codeWriter);
```

Comment:
This if-else block is for initializing `extendsTypes` and `implementsTypes`.

This actually belongs in the block above where we emit the name.

This implementation is incorrect if there are type variables, as we will emit the type variables before the record constructor.

Can we please add a test for a record with type variables?

Reply:
You want:
```
    record Taco<T>(List<T> name) {}
```
So you want to emit the header after `emitTypeVariables` I think right?

Reply:
Have written a test for this though